### ISSUE TITLE
Model for Materialize CDC data

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#2221]
## A model for recording Materialize change data capture

We have a variety of different places we read data from, and places we might want to write data to. I don't think we currently have a consistent take on what we expect to find in each of these places and what we should produce for each of them. I specifically mean the logical data there, rather than the serialization format or anything like that; what semantic information comes at us from our sources, and what do we need to provide back out at it?

This write-up means to propose a model for what we might "store" for change data capture, in the logical rather than physical sense. The goal is that if we all agree that this makes sense, then for each communication medium (Kafka, Kinesis, files, etc) there is a different question of how we shoe-horn this information into that format, and potentially how we impose that structure when it is not necessarily provided.

This probably subsumes #34, though we should sanity check that it addresses the issues raised there.

**tl;dr**: I propose we use the logical structure of differential dataflow "batches"

--- 

Differential dataflow batches are collections of update triples

    (data, time, diff)

where each collection also has three frontiers:

    lower, 
    upper,
    since,

Each of these is a timely dataflow frontier, but if it helps you can think of them as `Some(Timestamp)`, that is "either a timestamp, or no timestamp". They indicate either "timestamps greater or equal to this timestamp" or "no times" in the case of `None`.

A batch contains exactly the subset of updates `(data, time, diff)` its collection undergoes for which

    lower <= time
    !upper <= time

We typically see the batches as a sequence, where the `upper` of one batch is the `lower` of the next batch, and in this case the batches partition all updates, eventually revealing all changes, and communicating the inexorable advance of capabilities at the same time.

The `since` frontier is information about compaction, which we probably do not expect to see on a live stream of data, but we will want to maintain when we are stashing consolidated batches for other people; if we ship them off to S3 and then someone starts up a materialized session using that information, we'll want to pull in compacted batches rather than replaying the whole stream.

---

So, the proposal is that we should try and have materialized speak this lanugage to its inputs and outputs. Each different framework will require compatibility layers, but I believe it will help to have a specific target in mind, and then intentionally fill in the gaps between the models, with some sort of paste.

We have something close to this already with the BYO consistency and realtime timestamping, though they use different mechanisms. In each case, the updates themselves are found in the Kafka topics, and the bounding metadata are either received (in the BYO case) or produced automatically and persisted (in the RT case). 

At the same time, we don't have the full flexibility of batches here: records all land at the same time (either the lower or upper bound of the batch, depending on how it is interpreted); we have one metadata announcement per distinct time; we don't have anything like since compaction information (we're largely ignoring this at the moment). That doesn't have to be a bad thing, especially if we don't have any better information from the source, but I think we should be making these choices intentionally.

I propose that we consider bringing these two approaches together, not necessarily to the exclusion of existing techniques (but probably), with the specific goal of matching the framing and information of differential batches above.

---

This would give us guidance on how to handle Kafka sinks, to ensure that they are structured in a way that allows us to spin materialized back up after failure, but also structured in a way that if someone else spins up materialized and wants to read the same data it can (without needing a peek in the catalog of the first instance).

Materialize natively produces batches as outputs, and it should be relatively easy to yield data and batch metadata in whatever format we think is most appropriate for Kafka. That decision isn't bolted down, and should probably be responsive to compatibility requirements (e.g. that it should wrap existing streams with an additional metadata stream, rather than invent new punctuation).

This should also provide guidance on what shape Kinesis sources and sinks need to be hammered into to comport with all of this. Where there is space between what Kinesis supplies and what we need, we can build the parts that reproduce as if they were batches (perhaps similar metadata streams).

This should also provide guidance on what shape an S3 store would want to resemble, which would probably be wildly different than the streaming infrastructures (e.g. random access by `since`, and consolidated batches that can just be pulled down rather than re-read and re-built from a streaming source).

---

I think the main ask with this proposal is that other points in the system that have some local degree of autonomy might need to be wrangled in to line. Rather than produce timestamps on our own and write them just to the catalog, we might want to preserve them in a format that others would agree with (e.g. allow a second materialized instance to read the same Kafka stream and see the same updates).

I'm not 100% clear on what these points are or how painful they are, but I fairly strongly believe that we'll want to put a stake in the ground at some point with respect to how we think about the information we consume and produce. It seems like a great opportunity for some public-facing technical leadership too; if we do this well, we have the potential to bring other people along with us too.

I tried to think of downsides and I didn't come up with too many, other than the flexibility limitations. There is the possibility that we need more information than provided above, or that the framing above is too restrictive (for example: it doesn't support out of order timestamps; they have to present as in a batch and then `delay` themselves). But I figured posting this would stir up some thought.

**Comment by rjnn** [comment]
I agree with everything you said; no further points or questions. The questions you've raised are all the right ones, but I have no answers or additions that would help, you'll have to answer them yourself. From a product perspective, this is brilliant.

**Comment by cuongdo** [comment]
@frankmcsherry The points you brought up seem sound. I'm curious what next steps you envision. I'd also like to sketch out with this looks like in practice for the sources and sinks we support and will soon support. Let's discuss IRL.

**Comment by frankmcsherry** [comment]
Sketching out a migration for what we have seems like a sane next step. For example, a pretty brutal migration would be to recast the realtime timestamping as something that produces a BYO consistency topic (and perhaps some massaging to make that topic look like batches). This would then let us clearly spec out what a Kafka sink should do (something that is blocking @ruchirK as I understand it).

**Comment by nacrooks** [comment]
Trying to summarise what Frank and I were speaking about earlier today. Apologies if this is confusing. 

There are two options: 

1) use batches with explicit lower/upper timestamps that fully summarise the space of records that can be contained between the interval [lower,upper]. This means that the batch must contain everything that will ever be received between [lower,upper], but also cannot contain anything that is outside those intervals. 

Making that assumption makes recovery potentially easier (no need to track data that hasn't been "committed" yet) at the level of differential, but has the downside that it is not possible to process records "out of order".

It becomes "harder" to process records that arrive in this order:
![P_20200305_125728](https://user-images.githubusercontent.com/2697885/76010638-39e23500-5ee1-11ea-9302-4be147ee900a.jpg)

In this example, operations in T1 have timestamp 1, operations in T2 have timestamp 2, operations in T3 have timestamp 3.  Without buffering updates, there's never a "clean" way to structure batches so that they only contain records within the appropriate interval. The question is then: what do we do of these buffered updates in materialized, and what happens on failures. It seems to me that this approach "splits" the recovery problem to be part in materialized, part in differential, which seems suboptimal.

2) watermarks: watermarks simply close out timestamps (and can do so out of order). It removes the need to buffer updates (they can all be processed), but pushes the need to handle these out-of-order operations to diff dataflow (if I understood correctly) and differential doesn't speak watermarks other than through downgrading capabilities. My understanding is that right now materialize effectively provides timely with watermarks (by modulating when we call capability.downgrade), and sends records individually. Frank - what happens to a record with timestamp t-1 when it gets propagated to timely, but when we still have a capability for t-2?

3) multitemporal timestamps: we discussed a way to allow out of order processing of records in diff dataflow by having timestamps internally be bi-dimensional. This would allow us to "store" records as part of batches (creating a uniform framework for recovery), but without "exposing" these partial timestamps. The downside of this approach is that operations that are very far in the future would cause differential to have to do more computation work (as the space of possible "open" timestamp grows quadratically).

I know this is not Frank was thinking about when writing this, but thinking about it from a user's point of view, it seems to me that we want to guarantee two things
1) users shouldn't have to modify the ordering of their data-stream. If data arrives out of order, so be it (this is incidentally not something we support in Kafka but I think is going to be important to support in the future)
2) it shouldn't be hard for users to "compute" when to close out timestamps. Flink has these "watermark" events, I don't think we should make it any harder than that.

I don't think that API is particularly tied towards us choosing to use batches or watermarks as our internal representation.

**Comment by nacrooks** [comment]
> Sketching out a migration for what we have seems like a sane next step. For example, a pretty brutal migration would be to recast the realtime timestamping as something that produces a BYO consistency topic (and perhaps some massaging to make that topic look like batches). This would then let us clearly spec out what a Kafka sink should do (something that is blocking @ruchirK as I understand it).

The topic already looks like batches, so that wouldn't be a big issue. The only minor issue would be to modify materialize to support dataflow ids (as we discussed previously). Would it be necessary Frank to modify how we hand off individual records to Timely? (do we still have to do it individually or would we have to buffer them inside Materialize before handing them off once all the messages have arrived?)

**Comment by frankmcsherry** [comment]
> Frank - what happens to a record with timestamp t-1 when it gets propagated to timely, but when we still have a capability for t-2?

Timely will forward the record, and it should move through operators like Avro deserialization, map, filter, project, up until it reaches an `Arrange` operator, at which point it gets buffered until the frontier advances.

> The only minor issue would be to modify materialize to support dataflow ids (as we discussed previously).

I think one thing that might spill out of this is that we actually want each read of the same source topic to use the same consistency topic. I do think we probably want that, and that creating a new consistency topic would be a not-lightweight thing (doable, but probably not meant to be the default?). I may have forgotten what is hard and what is easy and why, though.

**Comment by nacrooks** [comment]
I'm not sure I follow your last comment.  Can you expand? 

It certainly would be possible for the first read of a source to write to a kafka/kinesis topic, and then subsequent reads of that source to default to reading from that consistency topic. We still (I suspect) want to preserve the ability of sources to read from *different* consistency topics

I can imagine (abusing syntax slightly) something like `CREATE MATERIAZED SOURCE from topic=x with (reproducibility=ON) `syntax. Subsequent calls of the form `CREATE MATERIALIZED SOURCE with (reproducability=ON) `would then check whether there is a consistency topic "x_reproducability", if yes, they would "default" to regular BYO consistency sources with that topic as the consistency topic, if not, they just start using RT consistency. The *big* difference with a "regular" consistency topic would be that, when we've finished processing that topic, we want to start writing to it again (so that we can fast-forward timestamps). 

One of the "downsides" of this approach is that it will assign timestamps that are consistent with the first read of the source, rather than the current read (until we finish processing the source and can fast forward the stream). This will create quite a lot of artificially "stale" result sets

**Comment by frankmcsherry** [comment]
Maybe a different way to say the same point: rather than "dataflow identifiers" I think we might want "consistency topics" to be the thing we write down about a source. It may turn out that the user wants to mint these all over the place, but it feels better to give them names and write them down and re-use them, rather than dataflow internal stuff.

That is independent from whether all sources should use the same consistency topic, but I wouldn't be surprised if we want to encourage that (otherwise, weird anomalies). One should be able to have multiple consistency topics for a data topic, for sure.

> The big difference with a "regular" consistency topic would be that, when we've finished processing that topic, we want to start writing to it again (so that we can fast-forward timestamps).

I'm not sure I understand this. It might related to #1210 though? My thinking was that if sources were created "as of" times, this gives the system permission to fast-forward using topics with compacted `since` frontiers. I'm not sure on what basis we would create these though (do we write to the consistency topic, or create a new one?).

> This will create quite a lot of artificially "stale" result sets

I think this is again related to #1210 which should allow us to bypass these stale result sets (e.g. the logic I imagine is that each change is fast-forwarded to the as-of time, even if they had lots of interesting different times when read from the input).

**Comment by frankmcsherry** [comment]
We talked a bit on Friday, and I wanted to record some notes here.

### 1. There are (at least) three qualitatively different persistence levels

Imagine that the (persistent) data stream we have coming in records both data and diffs (e.g. before, after), we may have to create (and maybe persist) timestamp and capability information. Specifically "at what time does each change occur?" and "at what point can we downgrade capabilities to allow computation to unfold"?

We think there are three tiers of persistence, each of which are legit but trade off the amount of prep work against the delightful properties provided by each. It may be sane to start each new framework with the simplest implementation, and build up to more complicated ones with time and experience with the framework.

The "delightful properties" mostly seem to relate to failure recovery and initial loading of data, rather than steady-state operation. The "amount of prep work" seems to be mostly in the cases where the input streams supply opinions about timestamps (if we can invent them ourselves, then things are much easier).

#### 1a. No persistence ("none")

Just maintain a monotonic system timer in each worker, use the time of receipt as the time for data, and downgrade the capability for this timer.

This sounds pretty rough, and doesn't have great persistence properties, but it does maintain an "eventual consistency" property that any replay will at least get the same changes applied, if not in exactly the same order. For many users this may be acceptable, and it seems like a not-unreasonable place to start, especially if the framework makes life difficult for the other approaches.

This approach makes failure recovery the hardest, as it is very difficult to correlate output we may have already produced (and which may have been read by someone) with the input. The best recovery strategy may just be to start over and create a new output stream (or use the same output stream, but subtract off its accumulated contents as the first action).

#### 1b. Persisted timestamps and capabilities ("watermarks")

When we do not have information about timestamps or capabilities, we may be able to create and then persist this information.

This approach is closest to what we do now, and provides the guarantee of deterministic replay: if we (or anyone else) starts over on the same stream, the same things will be produced with the same timestamps. For closed times (which capabilities have been downgraded past) the same conclusions will be reached, as subsequent (potentially unrecorded) decisions could not have produced output at these times yet (at least, as long as we "commit" output only once the input has been committed).

Failure recovery is still challenging in this case, as while there is a clearer connection between outputs and moments in the input (capability downgrades), the bound is only one-sided: we are sure no relevant input data exists after a capability downgrade, but it can exist arbitrarily far in the history of the input stream. Recovering a "failed" stream (e.g. an output that was not completely written to) to a consistent state may require re-scanning the whole stream to look for early records.

#### 1c. Framing "batches" a la differential ("batches")

We create and persist timestamp information in differential "batches", including lower and upper bounds on timestamps, and since bounds on the compaction frontier.

This approach is "hardest" in the sense that it is farthest from an undecorated input stream, and will probably require actual computation to produce. 

**Aside**: This computation doesn't have to be harder than differential's `arrange` operator: from a stream of the previous type, buffer all records until a capability downgrade, then produce (persist) a batch of buffered records not in advance of the new capability, and a batch of buffered records in advance of the new capability. The first batch can be published as a batch and read by downstream consumers, the second batch should be maintained as the "premature" records that we would want ready access to on recovery (vs re-scanning the input).

This approach does have the appealing property that we can most easily efficiently recover, as we have a clear map of where all records are. The lower and upper bounds also means that we can re-load data from compacted representations, by making sure that the upper and lower bounds line up, ensuring that we include each change exactly once.

### 2. Failure recover and re-loading data

For several reasons, we may want to load data using mechanisms other than attaching to an existing stream and playing it forward.

#### 2a. Failure recovery 

In the case of failure recovery, we may have a partly written output stream, whose contents cannot be retracted and which may have already been read by downstream consumers. To "recover" in this setting, output operators need to understand what has been produced into the stream, and how to splice on a new change stream seamlessly. It seems like this can be done in the "batch" case, where the output identifies existing batch boundaries in the output stream and in particular identifies the last complete batch: it can 1. read any other output updates and maintain them internally as retractions, and 2. suppress any outputs that would not be beyond the last completed batch. There may also need to be a 3. determine bounds on output updates to expand the bounds of the next output batch, so that it looks like a legit batch (differential's batching boundaries aren't guaranteed to align on replay).

#### 2b. Loading data

Although we do not currently do this, I believe that we will want to be able to load sources "as of" a time. For example, one may have a year's worth of historical changes, but only want to load a session "as of" now, or perhaps an hour ago, or some other time.

Loading a source "as of" a time means that 1. we could fast-forward all times to the as of bound, which will reduce the amount of computation done on loading, and (more importantly) 2. we can load compacted representations of the input data, bringing in something roughly equivalent to the accumulated representation of the data at the as of time.

To load compacted data, we need to represent the data in a way that makes it clear what we are pulling in. The batch representation does a good job of this, in that the lower and upper bounds allow the system to ensure that it has selected batches that disjointly cover the input times, and the since bound ensures that the data are correct for the as of time. 

Other representations make this more complicated, though perhaps not impossible. One could do Kafka compaction of a watermarked stream, and it could all work out well. It would start to become more complicated if we wanted to line up multiple frameworks, though: pulling in some compacted batches from S3 and attaching to an offset in Kafka for "live" updates.

### 3. Location / structure of metadata

Most of the schemes we've discussed are overlays, in the sense that we do not require the data stream to be rewritten. Even when we can re-write the stream (as in `materialized` outputs) we may not want to because of compatibility for downstream readers, and the ability to more efficiently scan metadata without scanning the whole data stream.

We do not need to couple the location of metadata and the framework. While we could write a metadata topic back in to Kafka (and that might make a lot of sense) it isn't mandatory, and there are reasons we may not want to do this (Kinesis streams cost to write, and so those users might prefer to stash the information elsewhere). We might want to pay attention to frameworks that would benefit from some exclusivity (perhaps some in-order guarantees or consistent "exactly once" guarantees only hold in-framework).

One example is to not put the metadata anywhere, corresponding to persistence tier 1c.

There is some benefit to putting the metadata somewhere public, rather than just in the catalog. Other users may want to use the same data, including other concurrent users of other materialized instances. If we just have the metadata in the catalog, this becomes harder (short of turning each materialized instance into a metadata server). Other places include the various frameworks themselves, which offer access control and persistence as their core competences.

It helps greatly if the metadata can be compact, roughly constant sized for each batch, in the batch persistence tier. In other tiers, we may be less likely to need to quickly look through the metadata, so perhaps the compactness is less important there as well.

### 4. Next steps

When creating a new framework, it would help everyone to establish some mapping between what the framework provides and some of the tiers of persistence. Clearly tier 1a is (we hope) pretty easy to do. It would be great to establish tier 1c as well as this is probably how we would like to produce data from differential; we will want to be able to read it back in to `materialized`. Tier 1b may also be helpful as an intermediate level, especially one that may not require any data manipulation/reorganization (though, I think 1c doesn't *require* data massaging, but does benefit from it).

Our approach to Kafka at the moment doesn't have a tier 1c framing, which is possibly blocking a fault-tolerant sink. It seems like it would also be blocking a hybrid S3 persistence story. We should probably design a tier 1c framing, if for no other reason than to check out what it looks like, and how complicated it is to produce from input data.

We should try and figure out if it is important to replace tier 1b streams with tier 1c information, or whether we can leave them as they are and just recover tier 1c efficiently when available. For example, it may be that we can compact tier 1b streams using Kafka offsets and compaction, which wouldn't generalize to other frameworks (e.g. Kinesis doesn't have offsets) but maybe that is fine and we don't need interoperable representations (maybe S3 is the only framework we need to hybridize with).

We should map out what fault-tolerant sinks for batched output looks like. @ruchirK and I started to talk about this, and it seems like one could recover tier 1c outputs with replay, potentially independent of the tiers of input to the computation. Should we ever produce anything other than tier 1c outputs? (i.e. should we always `consolidate` output? Should we produce out-of-order data when available?).

We should wire up `AS OF` for sources and start to use it. We aren't going to design things well if we don't exercise them. I'd like to do this ASAP, because there are immediate performance wins even without all of the framing stuff.

**Comment by elindsey** [comment]
I don't see anything actionable here, and most of this discussion has continued to percolate around and been incorporated into the latest designs around platform. Going to close this out
